File: hexwar/lua/hexwar/ui.lua
-- GUI 셸 — SPEC §11

local H = require('hexwar.hexcoord')
local hexmap = require('hexwar.hexmap')
local P = require('hexwar.path')
local PK = require('hexwar.picker')
local units = require('hexwar.units')

local M = {}

-- 위젯 종류
M.PANEL, M.BUTTON, M.LABEL, M.MINIMAP, M.MAPVIEW, M.LOG, M.DIALOG = 0, 1, 2, 3, 4, 5, 6
-- 상태 기계 (SPEC §11.2)
M.IDLE, M.SELECTED, M.TARGETING, M.DIALOG_ST, M.GAMEOVER = 0, 1, 2, 3, 4
M.STATE_NAMES = { [0] = 'IDLE', 'SELECTED', 'TARGETING', 'DIALOG', 'GAMEOVER' }

-- 화면 배치 상수는 렌더러와 공유한다
M.VIEW = { 0, 0, 256, 168 }
M.PANEL_R = { 256, 0, 64, 200 }
M.MSG = { 0, 168, 256, 32 }

local function widget(id, x, y, w, h, kind, label, children)
  return { id = id, x = x, y = y, w = w, h = h, kind = kind,
           label = label or '', enabled = true, visible = true, children = children or {} }
end
M.widget = widget

function M.build_ui()
  local minimap = widget('minimap', M.PANEL_R[1] + 6, 24, 48, 36, M.MINIMAP)
  local btn_end = widget('end', M.PANEL_R[1] + 4, 150, 56, 12, M.BUTTON, 'END TURN')
  local btn_undo = widget('undo', M.PANEL_R[1] + 4, 164, 56, 12, M.BUTTON, 'UNDO')
  local btn_next = widget('next', M.PANEL_R[1] + 4, 178, 56, 12, M.BUTTON, 'NEXT UNIT')
  local panel = widget('panel', M.PANEL_R[1], M.PANEL_R[2], M.PANEL_R[3], M.PANEL_R[4],
                       M.PANEL, '', { minimap, btn_end, btn_undo, btn_next })
  local mapview = widget('map', M.VIEW[1], M.VIEW[2], M.VIEW[3], M.VIEW[4], M.MAPVIEW)
  local logw = widget('log', M.MSG[1], M.MSG[2], M.MSG[3], M.MSG[4], M.LOG)
  local yes = widget('yes', 100, 112, 40, 14, M.BUTTON, 'YES')
  local no = widget('no', 180, 112, 40, 14, M.BUTTON, 'NO')
  local dlg = widget('dialog', 80, 74, 160, 56, M.DIALOG, 'END TURN?', { yes, no })
  dlg.visible = false
  return widget('root', 0, 0, 320, 200, M.PANEL, '', { mapview, logw, panel, dlg })
end

local function contains(w, px, py)
  return px >= w.x and px < w.x + w.w and py >= w.y and py < w.y + w.h
end

function M.hit_test(w, px, py)
  if not w.visible or not contains(w, px, py) then return nil end
  for i = #w.children, 1, -1 do
    local hit = M.hit_test(w.children[i], px, py)
    if hit then return hit end
  end
  if w.enabled then return w end
  return nil
end

local Ui = {}
Ui.__index = Ui
M.Ui = Ui

function M.new(g)
  local obj = {}
  for _, o in ipairs(g.objectives) do
    obj[g.map:axial_idx(o[1], o[2])] = true
  end
  return setmetatable({
    g = g, root = M.build_ui(), state = M.IDLE, prev_state = M.IDLE,
    cam_x = 0, cam_y = 0, cursor_idx = -1, sel_idx = -1, sel_unit = -1,
    move_overlay = {}, move_n = 0, attack_overlay = {}, attack_n = 0,
    reach = nil, objective_idx = obj,
  }, Ui)
end

function Ui:state_name() return M.STATE_NAMES[self.state] end

function Ui:ascii_log(g, n)
  local out = {}
  local total = #g.log
  for i = total, math.max(1, total - n + 1), -1 do
    out[#out + 1] = g.log[i][2]
  end
  return out
end

function Ui:clamp_cam()
  local render = require('hexwar.render')
  if self.cam_x < 0 then self.cam_x = 0 end
  if self.cam_y < 0 then self.cam_y = 0 end
  if self.cam_x > render.CAM_MAX_X then self.cam_x = render.CAM_MAX_X end
  if self.cam_y > render.CAM_MAX_Y then self.cam_y = render.CAM_MAX_Y end
end

function Ui:scroll(dx, dy)
  self.cam_x = self.cam_x + dx
  self.cam_y = self.cam_y + dy
  self:clamp_cam()
end

function Ui:center_on(idx)
  local row = idx // hexmap.MAP_W
  local col = idx - row * hexmap.MAP_W
  local cx, cy = PK.hex_center(col, row)
  self.cam_x = cx - M.VIEW[3] // 2
  self.cam_y = cy - M.VIEW[4] // 2
  self:clamp_cam()
end

function Ui:_attack_targets(u)
  local out, n = {}, 0
  if u.ammo <= 0 or u.mp <= 0 then return out, n end
  local m = self.g.map
  for _, tid in ipairs(self.g.pool:alive_ids()) do
    local t = self.g.pool:get(tid)
    if t.side ~= u.side then
      local i = m:axial_idx(t.q, t.r)
      if i >= 0 and m.fog[i] == hexmap.FOG_VISIBLE
          and H.distance(u.q, u.r, t.q, t.r) <= units.K_RNG[u.kind] then
        out[i] = true
        n = n + 1
      end
    end
  end
  return out, n
end

function Ui:select(uid)
  local u = self.g.pool:get(uid)
  if not u or u.side ~= self.g.side then return false end
  self.sel_unit = uid
  self.sel_idx = self.g.map:axial_idx(u.q, u.r)
  self.reach = P.reachable(self.g.map, self.g.pool, u)
  local mv, n = {}, 0
  for _, i in ipairs(self.reach.list) do
    if i ~= self.sel_idx and self.g.map.occupant[i] == units.NO_UNIT then
      mv[i] = true
      n = n + 1
    end
  end
  self.move_overlay, self.move_n = mv, n
  self.attack_overlay, self.attack_n = self:_attack_targets(u)
  self.state = M.SELECTED
  return true
end

function Ui:deselect()
  self.sel_unit = -1
  self.sel_idx = -1
  self.reach = nil
  self.move_overlay, self.move_n = {}, 0
  self.attack_overlay, self.attack_n = {}, 0
  self.state = M.IDLE
end

function Ui:next_unit()
  local ids = {}
  for _, uid in ipairs(self.g.pool:alive_ids(self.g.side)) do
    if self.g.pool:get(uid).mp > 0 then ids[#ids + 1] = uid end
  end
  if #ids == 0 then return false end
  local nxt = ids[1]
  for k, v in ipairs(ids) do
    if v == self.sel_unit then
      nxt = ids[(k % #ids) + 1]
      break
    end
  end
  if self:select(nxt) then
    self:center_on(self.sel_idx)
    return true
  end
  return false
end

function Ui:handle(ev)
  local kind, a, b = ev:match('^(%a+)%s*(%-?%w*)%s*(%-?%w*)$')
  if kind == 'click' then return self:on_click(tonumber(a), tonumber(b)) end
  if kind == 'key' then return self:on_key(a) end
  if kind == 'render' then return true end
  error('알 수 없는 이벤트: ' .. ev)
end

function Ui:on_click(x, y)
  local w = M.hit_test(self.root, x, y)
  if not w then return false end
  if self.state == M.DIALOG_ST then
    if w.id == 'yes' then
      self:close_dialog()
      self.g:end_turn()
      self:after_turn()
      return true
    elseif w.id == 'no' then
      self:close_dialog()
      return true
    end
    return false
  end
  if w.kind == M.BUTTON then return self:on_button(w.id) end
  if w.id == 'minimap' then return self:on_minimap(x, y, w) end
  if w.kind == M.MAPVIEW then return self:on_map_click(x, y) end
  return false
end

function Ui:on_button(wid)
  if wid == 'end' then
    self:open_dialog()
    return true
  elseif wid == 'undo' then
    local ok = self.g:undo()
    if ok and self.sel_unit >= 0 then
      if not self.g.pool:get(self.sel_unit) then
        self:deselect()
      else
        self:select(self.sel_unit)
      end
    end
    return ok
  elseif wid == 'next' then
    return self:next_unit()
  end
  return false
end

function Ui:on_minimap(x, y, w)
  local col = math.max(0, math.min(hexmap.MAP_W - 1, (x - w.x) // 2))
  local row = math.max(0, math.min(hexmap.MAP_H - 1, (y - w.y) // 2))
  self:center_on(row * hexmap.MAP_W + col)
  return true
end

function Ui:on_map_click(x, y)
  local col, row = PK.pick(x, y, self.cam_x, self.cam_y)
  if col == nil then return false end
  local i = row * hexmap.MAP_W + col
  self.cursor_idx = i
  local m = self.g.map
  local uid = m.occupant[i]

  if self.state == M.TARGETING then
    if self.attack_overlay[i] and uid ~= units.NO_UNIT then
      self.g:attack(self.sel_unit, uid)
      self:after_action()
      return true
    end
    self.state = M.SELECTED
    return false
  end

  if self.state == M.SELECTED then
    if self.attack_overlay[i] and uid ~= units.NO_UNIT then
      self.g:attack(self.sel_unit, uid)
      self:after_action()
      return true
    end
    if self.move_overlay[i] then
      self.g:move_unit(self.sel_unit, i)
      self:after_action()
      return true
    end
  end

  if uid ~= units.NO_UNIT and m.fog[i] == hexmap.FOG_VISIBLE then
    local u = self.g.pool:get(uid)
    if u and u.side == self.g.side then return self:select(uid) end
  end
  self:deselect()
  return true
end

function Ui:on_key(k)
  if self.state == M.DIALOG_ST then
    if k == 'ESC' then
      self:close_dialog()
      return true
    elseif k == 'ENTER' then
      self:close_dialog()
      self.g:end_turn()
      self:after_turn()
      return true
    end
    return false
  end
  if k == 'LEFT' then self:scroll(-PK.HEX_W, 0)
  elseif k == 'RIGHT' then self:scroll(PK.HEX_W, 0)
  elseif k == 'UP' then self:scroll(0, -PK.ROW_STEP)
  elseif k == 'DOWN' then self:scroll(0, PK.ROW_STEP)
  elseif k == 'TAB' then return self:next_unit()
  elseif k == 'U' then return self:on_button('undo')
  elseif k == 'E' then self:open_dialog()
  elseif k == 'T' then
    if self.state == M.SELECTED and self.attack_n > 0 then
      self.state = M.TARGETING
    else
      return false
    end
  elseif k == 'ESC' then
    if self.state == M.TARGETING then
      self.state = M.SELECTED
    else
      self:deselect()
    end
  else
    return false
  end
  return true
end

function Ui:_dlg()
  for _, c in ipairs(self.root.children) do
    if c.id == 'dialog' then return c end
  end
  error('dialog')
end

function Ui:open_dialog()
  self.prev_state = self.state
  self.state = M.DIALOG_ST
  self:_dlg().visible = true
end

function Ui:close_dialog()
  self:_dlg().visible = false
  self.state = self.prev_state
end

function Ui:after_action()
  local u = self.g.pool:get(self.sel_unit)
  if self.g.over then
    self.state = M.GAMEOVER
    self.move_overlay, self.move_n = {}, 0
    self.attack_overlay, self.attack_n = {}, 0
    return
  end
  if not u or (u.mp <= 0 and u.ammo <= 0) then
    self:deselect()
  else
    self:select(self.sel_unit)
  end
end

function Ui:after_turn()
  self:deselect()
  if self.g.over then self.state = M.GAMEOVER end
end

function Ui:digest()
  return string.format('%s|sel=%d|cur=%d|cam=%d,%d|mov=%d|atk=%d',
    self:state_name(), self.sel_unit, self.cursor_idx,
    self.cam_x, self.cam_y, self.move_n, self.attack_n)
end

return M

File: hexwar/py/hexwar/ui.py
# -*- coding: utf-8 -*-
"""GUI 셸 — SPEC §11.

   도스 게임의 GUI 는 툴킷이 없었다. 위젯 트리를 직접 만들고, 마우스 위치로
   히트 테스트를 하고, 상태 기계로 '지금 클릭이 무슨 뜻인지'를 정했다.
   여기 있는 세 조각이 그때 GUI 코드의 전부라고 봐도 된다.

     1. 위젯 트리 — 사각형과 부모/자식. 그리기 순서 = 히트 테스트 역순.
     2. 상태 기계 — 같은 클릭이 상태에 따라 선택·이동·공격이 된다.
     3. 명령 스택 — 상태를 바꾸는 길이 하나뿐이어야 언두가 성립한다.
"""

from . import los
from . import path as P
from . import picker as PK
from .hexmap import FOG_VISIBLE, MAP_W
from .render import CAM_MAX_X, CAM_MAX_Y, MSG, PANEL, VIEW
from .units import K_RNG, NO_UNIT

# 위젯 종류
PANEL_W, BUTTON, LABEL, MINIMAP, MAPVIEW, LOG, DIALOG = range(7)

# 상태 기계 (SPEC §11.2)
IDLE, SELECTED, TARGETING, DIALOG_ST, GAMEOVER = range(5)
STATE_NAMES = ('IDLE', 'SELECTED', 'TARGETING', 'DIALOG', 'GAMEOVER')


class Widget(object):
    __slots__ = ('id', 'x', 'y', 'w', 'h', 'kind', 'label', 'enabled', 'visible', 'children')

    def __init__(self, wid, x, y, w, h, kind, label='', children=None):
        self.id = wid
        self.x, self.y, self.w, self.h = x, y, w, h
        self.kind = kind
        self.label = label
        self.enabled = True
        self.visible = True
        self.children = children or []

    def contains(self, px, py):
        return self.x <= px < self.x + self.w and self.y <= py < self.y + self.h


def build_ui():
    """화면 배치는 코드에 박아 둔다. 리소스 파일로 뺄 수도 있지만, 320x200
       한 해상도만 지원하던 시절에는 이렇게 두는 편이 흔했다."""
    minimap = Widget('minimap', PANEL[0] + 6, 24, 48, 36, MINIMAP)
    btn_end = Widget('end', PANEL[0] + 4, 150, 56, 12, BUTTON, 'END TURN')
    btn_undo = Widget('undo', PANEL[0] + 4, 164, 56, 12, BUTTON, 'UNDO')
    btn_next = Widget('next', PANEL[0] + 4, 178, 56, 12, BUTTON, 'NEXT UNIT')
    panel = Widget('panel', PANEL[0], PANEL[1], PANEL[2], PANEL[3], PANEL_W, '',
                   [minimap, btn_end, btn_undo, btn_next])
    mapview = Widget('map', VIEW[0], VIEW[1], VIEW[2], VIEW[3], MAPVIEW)
    logw = Widget('log', MSG[0], MSG[1], MSG[2], MSG[3], LOG)
    yes = Widget('yes', 100, 112, 40, 14, BUTTON, 'YES')
    no = Widget('no', 180, 112, 40, 14, BUTTON, 'NO')
    dlg = Widget('dialog', 80, 74, 160, 56, DIALOG, 'END TURN?', [yes, no])
    dlg.visible = False
    root = Widget('root', 0, 0, 320, 200, PANEL_W, '', [mapview, logw, panel, dlg])
    return root


def hit_test(w, px, py):
    """뒤에서 앞으로 훑어 가장 위의 위젯을 찾는다. 자식이 부모보다 위다.
       모달 대화상자가 아래를 가리는 것도 이 순서 하나로 해결된다."""
    if not w.visible or not w.contains(px, py):
        return None
    for c in reversed(w.children):
        hit = hit_test(c, px, py)
        if hit is not None:
            return hit
    return w if w.enabled else None


class Ui(object):
    __slots__ = ('g', 'root', 'state', 'cam_x', 'cam_y', 'cursor_idx', 'sel_idx',
                 'sel_unit', 'move_overlay', 'attack_overlay', 'reach', 'prev_state',
                 'dialog', 'msg', 'objective_idx')

    def __init__(self, g):
        self.g = g
        self.root = build_ui()
        self.state = IDLE
        self.prev_state = IDLE
        self.cam_x = 0
        self.cam_y = 0
        self.cursor_idx = -1
        self.sel_idx = -1
        self.sel_unit = -1
        self.move_overlay = set()
        self.attack_overlay = set()
        self.reach = {}
        self.dialog = None
        self.msg = []
        self.objective_idx = set(g.map.axial_idx(q, r) for (q, r) in g.objectives)

    # ------------------------------------------------------------- 표시용
    def state_name(self):
        return STATE_NAMES[self.state]

    def ascii_log(self, g, n):
        rows = [a for (_k, a) in g.log[-n:]]
        return list(reversed(rows))

    # --------------------------------------------------------------- 카메라
    def clamp_cam(self):
        self.cam_x = max(0, min(CAM_MAX_X, self.cam_x))
        self.cam_y = max(0, min(CAM_MAX_Y, self.cam_y))

    def scroll(self, dx, dy):
        self.cam_x += dx
        self.cam_y += dy
        self.clamp_cam()

    def center_on(self, idx):
        """헥스가 화면 가운데 오도록. 맵 끝에서는 잘리지 않게 물린다."""
        row, col = divmod(idx, MAP_W)
        cx, cy = PK.hex_center(col, row)
        self.cam_x = cx - VIEW[2] // 2
        self.cam_y = cy - VIEW[3] // 2
        self.clamp_cam()

    # --------------------------------------------------------------- 선택
    def select(self, uid):
        u = self.g.pool.get(uid)
        if u is None or u.side != self.g.side:
            return False
        self.sel_unit = uid
        self.sel_idx = self.g.map.axial_idx(u.q, u.r)
        self.reach = P.reachable(self.g.map, self.g.pool, u)
        self.move_overlay = set(i for i in self.reach.list
                                if i != self.sel_idx and self.g.map.occupant[i] == NO_UNIT)
        self.attack_overlay = self._attack_targets(u)
        self.state = SELECTED
        return True

    def _attack_targets(self, u):
        """사거리 안에 있고, 보이고, 상대편인 유닛이 선 칸."""
        out = set()
        if u.ammo <= 0 or u.mp <= 0:
            return out
        m = self.g.map
        for t in self.g.pool.iter_alive():
            if t.side == u.side:
                continue
            i = m.axial_idx(t.q, t.r)
            if i < 0 or m.fog[i] != FOG_VISIBLE:
                continue
            from . import hexcoord as H
            if H.distance(u.q, u.r, t.q, t.r) <= K_RNG[u.kind]:
                out.add(i)
        return out

    def deselect(self):
        self.sel_unit = -1
        self.sel_idx = -1
        self.reach = {}
        self.move_overlay = set()
        self.attack_overlay = set()
        self.state = IDLE

    def next_unit(self):
        """이동력이 남은 아군을 순서대로 돈다. 도스 워게임의 관용 조작."""
        ids = [u.id for u in self.g.pool.iter_alive(self.g.side) if u.mp > 0]
        if not ids:
            return False
        if self.sel_unit in ids:
            nxt = ids[(ids.index(self.sel_unit) + 1) % len(ids)]
        else:
            nxt = ids[0]
        if self.select(nxt):
            self.center_on(self.sel_idx)
            return True
        return False

    # ---------------------------------------------------------------- 입력
    def handle(self, ev):
        """이벤트 문자열 하나를 처리한다. 골든 트레이스가 이 문자열을 재생한다."""
        parts = ev.split()
        kind = parts[0]
        if kind == 'click':
            return self.on_click(int(parts[1]), int(parts[2]))
        if kind == 'key':
            return self.on_key(parts[1])
        if kind == 'render':
            return True
        raise ValueError('알 수 없는 이벤트: %r' % ev)

    def on_click(self, x, y):
        w = hit_test(self.root, x, y)
        if w is None:
            return False
        if self.state == DIALOG_ST:
            # 모달: 대화상자 밖의 클릭은 통째로 버린다
            if w.id == 'yes':
                self.close_dialog()
                self.g.end_turn()
                self.after_turn()
                return True
            if w.id == 'no':
                self.close_dialog()
                return True
            return False
        if w.kind == BUTTON:
            return self.on_button(w.id)
        if w.id == 'minimap':
            return self.on_minimap(x, y, w)
        if w.kind == MAPVIEW:
            return self.on_map_click(x, y)
        return False

    def on_button(self, wid):
        if wid == 'end':
            self.open_dialog()
            return True
        if wid == 'undo':
            ok = self.g.undo()
            if ok and self.sel_unit >= 0:
                if self.g.pool.get(self.sel_unit) is None:
                    self.deselect()
                else:
                    self.select(self.sel_unit)
            return ok
        if wid == 'next':
            return self.next_unit()
        return False

    def on_minimap(self, x, y, w):
        """미니맵 클릭 = 그 자리로 카메라 이동. 2픽셀이 헥스 하나다."""
        col = min(MAP_W - 1, max(0, (x - w.x) // 2))
        row = min(17, max(0, (y - w.y) // 2))
        self.center_on(row * MAP_W + col)
        return True

    def on_map_click(self, x, y):
        hexpos = PK.pick(x, y, self.cam_x, self.cam_y)
        if hexpos is None:
            return False
        col, row = hexpos
        i = row * MAP_W + col
        self.cursor_idx = i
        m = self.g.map
        uid = m.occupant[i]

        if self.state == TARGETING:
            if i in self.attack_overlay and uid != NO_UNIT:
                self.g.attack(self.sel_unit, uid)
                self.after_action()
                return True
            self.state = SELECTED
            return False

        if self.state == SELECTED:
            if i in self.attack_overlay and uid != NO_UNIT:
                self.g.attack(self.sel_unit, uid)
                self.after_action()
                return True
            if i in self.move_overlay:
                self.g.move_unit(self.sel_unit, i)
                self.after_action()
                return True

        if uid != NO_UNIT and m.fog[i] == FOG_VISIBLE:
            u = self.g.pool.get(uid)
            if u is not None and u.side == self.g.side:
                return self.select(uid)
        self.deselect()
        return True

    def on_key(self, k):
        if self.state == DIALOG_ST:
            if k == 'ESC':
                self.close_dialog()
                return True
            if k == 'ENTER':
                self.close_dialog()
                self.g.end_turn()
                self.after_turn()
                return True
            return False
        if k == 'LEFT':
            self.scroll(-PK.HEX_W, 0)
        elif k == 'RIGHT':
            self.scroll(PK.HEX_W, 0)
        elif k == 'UP':
            self.scroll(0, -PK.ROW_STEP)
        elif k == 'DOWN':
            self.scroll(0, PK.ROW_STEP)
        elif k == 'TAB':
            return self.next_unit()
        elif k == 'U':
            return self.on_button('undo')
        elif k == 'E':
            self.open_dialog()
        elif k == 'T':
            if self.state == SELECTED and self.attack_overlay:
                self.state = TARGETING
            else:
                return False
        elif k == 'ESC':
            if self.state == TARGETING:
                self.state = SELECTED
            else:
                self.deselect()
        else:
            return False
        return True

    # ------------------------------------------------------------ 대화상자
    def open_dialog(self):
        self.prev_state = self.state
        self.state = DIALOG_ST
        self._dlg().visible = True

    def close_dialog(self):
        self._dlg().visible = False
        self.state = self.prev_state

    def _dlg(self):
        for c in self.root.children:
            if c.id == 'dialog':
                return c
        raise KeyError('dialog')

    # -------------------------------------------------------------- 뒤처리
    def after_action(self):
        u = self.g.pool.get(self.sel_unit)
        if self.g.over:
            self.state = GAMEOVER
            self.move_overlay = set()
            self.attack_overlay = set()
            return
        if u is None or (u.mp <= 0 and u.ammo <= 0):
            self.deselect()
        else:
            self.select(self.sel_unit)

    def after_turn(self):
        self.deselect()
        if self.g.over:
            self.state = GAMEOVER

    # -------------------------------------------------------------- 요약값
    def digest(self):
        return '%s|sel=%d|cur=%d|cam=%d,%d|mov=%d|atk=%d' % (
            self.state_name(), self.sel_unit, self.cursor_idx,
            self.cam_x, self.cam_y, len(self.move_overlay), len(self.attack_overlay))
